library/alloc/src/str.rs

```diff
@@ -280,8 +279,16 @@ impl str {
                 return unsafe { replace_ascii(self.as_bytes(), from_byte, *to_byte) };
             }
         }
-
-        let mut result = String::new();
+        let to_len = to.as_bytes().len();
+        // Set result capacity to self.len() when from.len() <= to.len()
+        let default_capacity = match from.as_utf8_pattern() {
+            Some(Utf8Pattern::StringPattern(s)) if s.len() <= to_len => self.len(),
+            Some(Utf8Pattern::CharPattern(c)) if c.encode_utf8(&mut [0; 4]).len() <= to_len => {
```

Comment:
There's a dedicated method for this:
```suggestion
            Some(Utf8Pattern::CharPattern(c)) if c.len_utf8() <= to_len => {
```